### PR TITLE
Improve CLI managementment

### PR DESCRIPTION
**Pull request by ZidongLiu**
# Reference

https://docs.pydantic.dev/latest/concepts/pydantic_settings/#command-line-support

# Task

- [x] remove checkpoint
- [x] Use pydantic to define cli
- [x] Cli to Python experiment components
- [x] Move preprocess data
- [x] Remove config.py, and all related changes
- [x] Documentation for Cli
- [x] unit test parser
- [x] test training process

# Notes:
Mix like this way can not work.

image

### CODE REVIEW

**Comment by ZidongLiu** [comment]
decomfl:
Squad trains: `python decomfl_main.py --large-model=opt-125m --dataset=squad --iterations=1000 --train-batch-size=8 --test-batch-size=8 --eval-iterations=25 --num-clients=1 --num-sample-clients=1 --local-update-steps=1 --num-pert=2 --lr=1e-5 --mu=1e-3 --grad-estimate-method=rge-forward`

image

sst2 works: `python decomfl_main.py --large-model=opt-125m --dataset=sst2 --iterations=1000 --train-batch-size=32 --test-batch-size=200 --eval-iterations=25 --num-clients=3 --num-sample-clients=2 --local-update-steps=1 --num-pert=5 --lr=1e-5 --mu=1e-3 --grad-estimate-method=rge-forward`

image

---

zo_rge_main:

mnist trains to 60% after 2 epochs: `python zo_rge_main.py --dataset=mnist --num-pert=10 --lr=1e-5 --mu=1e-3 --momentum=0.9`

image

---
fo_fl_main:

mnist can train to 85% in just few steps: `python fo_fl_main.py --dataset=mnist --lr=1e-3 --num-clients=5 --num-sample-clients=3 --local-update-steps=1 --train-batch-size=256 --test-batch-size=1000 --momentum=0.9`

image

sst2 can be train to 86%: `python fo_fl_main.py --dataset=sst2 --lr=1e-3 --num-clients=5 --num-sample-clients=3 --local-update-steps=1 --train-batch-size=32 --test-batch-size=200 --momentum=0.9`

image

---
llm_fine_tune works for 

sst2, `python llm_fo_fine_tune_main.py --dataset=sst2`

image

File: decomfl_main.py [hunk 1]
```diff
@@ -9,31 +9,57 @@
...
+    GeneralSetting,
+    DeviceSetting,
+    DataSetting,
+    ModelSetting,
+    OptimizerSetting,
+    FederatedLearningSetting,
+    RGESetting,
+    ByzantineSetting,
+):
+    pass
```
Comment (BichengYing) [inline decomfl_main.py:40]: add comment about this class since this is not obvious at the first look

File: decomfl_main.py [hunk 2]
```diff
@@ -9,31 +9,57 @@
...
-        client_model.to(client_device)
+        client_model = prepare_settings.get_model(
+            dataset=args.dataset, model_setting=args, seed=args.seed
+        ).to(client_device)
+        client_optimizer = prepare_settings.get_optimizer(
+            model=client_model, dataset=args.dataset, optimizer_setting=args
+        )
+        client_grad_estimator = prepare_settings.get_random_gradient_estimator(
+            model=client_model, device=client_device, rge_setting=args, model_setting=args
+        )
```
Comment (BichengYing) [inline decomfl_main.py:62]: I like this style but the drawback is this `model_setting=args`, `optimizer_setting=args`, `rge_setting=args`....

The args is not really split but by setting type but passing it repeatedly and using duck typing style idea.

File: experiment_helper/cli_parser.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+from experiment_helper.experiment_typing import (
+    LargeModel,
+    ModelDtype,
+)
+from cezo_fl.random_gradient_estimator import RandomGradEstimateMethod
+from cezo_fl.util.language_utils import SUPPORTED_LLM
+from experiment_helper.data import DataSetting  # noqa: F401
+
+
+class GeneralSetting(BaseSettings, cli_parse_args=True):
```
Comment (BichengYing) [inline experiment_helper/cli_parser.py:15]: is there an easy to list all flags?

File: experiment_helper/data.py [hunk 4]
```diff
@@ -1,66 +1,56 @@
...
-        server_device = {"server": torch.device("cpu")}
-        client_devices = {get_client_name(i): torch.device("cpu") for i in range(num_clients)}
+class ImageClassificationTask(Enum):
+    mnist = "mnist"
+    cifar10 = "cifar10"
+    fashion = "fashion"
+
 
-    return server_device | client_devices, kwargs
+class DataSetting(BaseSettings, cli_parse_args=True):
```
Comment (BichengYing) [inline experiment_helper/data.py:32]: so we may have multiple flags that do not come from the same file, right?

File: experiment_helper/cli_parser.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import torch
+from pydantic import Field, AliasChoices
+from pydantic_settings import BaseSettings, CliImplicitFlag
+from experiment_helper.experiment_typing import (
+    LargeModel,
+    ModelDtype,
+)
+from cezo_fl.random_gradient_estimator import RandomGradEstimateMethod
+from cezo_fl.util.language_utils import SUPPORTED_LLM
+from experiment_helper.data import DataSetting  # noqa: F401
```
[the remark this answers is not in the document]
Comment (ZidongLiu) [reply experiment_helper/cli_parser.py:12]: I import this for a central place to export. I want to make this cli_parser the only place other module can access the defined parser

File: decomfl_main.py (hunk 2 above)
Comment (ZidongLiu) [reply decomfl_main.py]: See my latest commit for some code design idea. [try to resolve comments, may revert](https://github.com/ZidongLiu/DeComFL/pull/73/commits/5ed9088aa39cb39f06a7d1bfea544828a9c9084e)

I tried to make something like this, but I got some issue with the usage and unit test. we need to chat about it.

File: experiment_helper/cli_parser.py (hunk 3 above)
Comment (ZidongLiu) [reply experiment_helper/cli_parser.py]: I do not fully understand the question

Comment (ZidongLiu) [reply experiment_helper/cli_parser.py]: we want to have a list to tell how many settings there are?

something like this?:
[GeneralSetting, DeviceSetting, ModelSetting, ...]

File: experiment_helper/data.py (hunk 4 above)
Comment (ZidongLiu) [reply experiment_helper/data.py:32]: this one is special, I am also not sure if we should move this one to the central cli_parser.py. 

I tried to group related logic into same file as much as possible. 
Originally, I wanted to define `class ImageClassificationTask(Enum)` and `get_dataloaders` in this file, and define `DataSetting` in cli_parser.py. But got circular dependency. So I have to choose between 2 solutions:

1. Define `class ImageClassificationTask(Enum)` in a separate file, probably `experiment_typing.py`. Then use it in 2 files.
2. Define everything in this `data.py` file, and import DataSetting in `cli_parser.py` which is what I am doing

File: decomfl_main.py (hunk 1 above)
Comment (ZidongLiu) [reply decomfl_main.py]: will add in the end after we resolve the other comments

**Comment by ZidongLiu** [comment]
using cached property so that created object can be reused

image

File: experiment_helper/cli_parser.py (hunk 3 above)
Comment (ZidongLiu) [reply experiment_helper/cli_parser.py:15]: Will add some docstring in the future when we are going to make this repo a library, need to give some documentation for user

File: decomfl_main.py (hunk 2 above)
Comment (ZidongLiu) [reply decomfl_main.py:62]: resolved by adding @cache_property to all Settings

**Comment by ZidongLiu** [comment]
@BichengYing ready for review again

**Comment by ZidongLiu** [comment]
Added frozen to all setting classes. All instance should be immutable.

image

**Comment by ZidongLiu** [comment]
decomfl works.

image

**Pull request merged**